lib/calendar.ts

```diff
@@ -1852,170 +1818,170 @@ function isGregorianLeapYear(year: number) {
 }
 
 /** Base for all Gregorian-like calendars. */
-const makeHelperGregorian = (id: BuiltinCalendarId, originalEras: InputEra[]) => {
-  const { eras, anchorEra } = adjustEras(originalEras);
-  const helperGregorian = ObjectAssign({}, helperSharedImpl, {
-    id,
-    eras,
-    anchorEra,
-    calendarType: 'solar',
-    inLeapYear(this: HelperPerCalendarImpl, calendarDate: CalendarYearOnly /*, cache: OneObjectCache */) {
-      // Calendars that don't override this method use the same months and leap
-      // years as Gregorian. Once we know the ISO year corresponding to the
-      // calendar year, we'll know if it's a leap year or not.
-      const { year } = this.estimateIsoDate({ month: 1, day: 1, year: calendarDate.year });
-      return isGregorianLeapYear(year);
-    },
-    monthsInYear(/* calendarDate */) {
-      return 12;
-    },
-    minimumMonthLength(this: HelperPerCalendarImpl, calendarDate: CalendarYM) {
-      const { month } = calendarDate;
-      if (month === 2) return this.inLeapYear(calendarDate) ? 29 : 28;
-      return [4, 6, 9, 11].indexOf(month) >= 0 ? 30 : 31;
-    },
-    maximumMonthLength(this: HelperPerCalendarImpl, calendarDate: CalendarYM) {
-      return this.minimumMonthLength(calendarDate);
-    },
-    /** Fill in missing parts of the (year, era, eraYear) tuple */
-    completeEraYear(this: HelperPerCalendarImpl & { eras: Era[] }, calendarDate: Partial<FullCalendarDate>) {
-      const checkField = (name: keyof FullCalendarDate, value: string | number | undefined) => {
-        const currentValue = calendarDate[name];
-        if (currentValue != null && currentValue != value) {
-          throw new RangeError(`Input ${name} ${currentValue} doesn't match calculated value ${value}`);
-        }
-      };
-      const eraFromYear = (year: number) => {
-        let eraYear;
-        const adjustedCalendarDate = { ...calendarDate, year };
-        const matchingEra = this.eras.find((e, i) => {
-          if (i === this.eras.length - 1) {
-            if (e.reverseOf) {
-              // This is a reverse-sign era (like BCE) which must be the oldest
-              // era. Count years backwards.
-              if (year > 0) throw new RangeError(`Signed year ${year} is invalid for era ${e.name}`);
-              eraYear = e.anchorEpoch.year - year;
-              return true;
-            }
-            // last era always gets all "leftover" (older than epoch) years,
-            // so no need for a comparison like below.
-            eraYear = year - e.anchorEpoch.year + (e.hasYearZero ? 0 : 1);
-            return true;
-          }
-          const comparison = helperSharedImpl.compareCalendarDates(adjustedCalendarDate, e.anchorEpoch);
-          if (comparison >= 0) {
-            eraYear = year - e.anchorEpoch.year + (e.hasYearZero ? 0 : 1);
+abstract class GregorianBaseHelper extends HelperBase {
+  id: BuiltinCalendarId;
```

Comment:
Doesn't this need to be abstract to force implementations to specify it?

Reply:
TS wasn't happy about making these props `abstract`, because of this error: 

> Abstract property 'id' in class 'GregorianBaseHelper' cannot be accessed in the constructor.

I guess if properties are assigned in the constructor, then they can't be abstract.

Reply:
ah - I didn't notice the duplication there. Would it be worth making it a readonly constructor property instead of specifying it twice?

```
constructor(readonly id: BuildinCalendarId, originalEras: InputEra[]) {
	super();
	// no need for this.id = id;
	// or adding the property
}
```

Reply:
If all the constructor params were setting properties without code, sure that'd work. But the `originalEras` parameter is split into two properties by the `adjustEras` function.  I'm not sure it makes things any more readable to only use the `readonly` trick for `id` but then run code for the era-related properties. Seems clearer to me to just explicitly show what's happening. What do you think?

Reply:
This is the last outstanding issue before merging and I don't think there's a clear obvious answer either way, so I'm planning to leave as-is for now. Happy to revisit later.